gnmic: add page and subpages (#8367)
tldr-pages/tldr: pages/common/gnmic.md

# gnmic

> > A gNMI CLI client that provides full support for Capabilities, Get, Set and Subscribe RPCs with collector capabilities.
> > More information: <https://gnmic.kmrd.dev>.

Reviewer: Taking the perspective of someone that's not familiar with `gnmic` or gNMI more broadly, I'd want this description to explain what `gnmic` is used for and why I might use it. This description indicates what gnmic is capable of but if I'm a new user that just encountered this command in a script and want to figure out what it's doing, I won't gain that understanding beyond knowing that I need to start searching for "gNMI".

This [Cisco gNMI Protocol](https://www.cisco.com/c/en/us/td/docs/ios-xml/ios/prog/configuration/168/b_168_programmability_cg/gNMI_protocol.pdf) document indicates that gNMI is used to interact with/set configuration on network devices. Would the following be an accurate description of what gnmic is used for?

```suggestion
> A gNMI command-line client.
> Manage gNMI network device configuration and view operational data.
> More information: <https://gnmic.kmrd.dev>.
```

> - Request capabilities (gnmic capabilities):

Reviewer: ```suggestion
- Request device capabilities:
```

`gnmic --address {{ip:port}} --insecure capabilities`

> - Username and password can be given to any command:

Reviewer: Command examples typically describe what a command is doing and how it's used, can we indicate 

```suggestion
- Provide a username and password to fetch device capabilities:
```

`gnmic --address {{ip:port}} --username {{username}} --password {{password}} --insecure capabilities`

> - Get request (gnmic get):

Reviewer: ```suggestion
- Get a snapshot of the device state at {{path}}:
```

`gnmic -a {{ip:port}} --insecure get --path {{path}}`

> - Set request (gnmic set):

Reviewer: Going off if the [gNMI specification section 3.4](https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-specification.md#34-modifying-state)

```suggestion
- Update device state at {{path}}:
```

`gnmic -a {{ip:port}} --insecure set --update-path {{path}} --update-value {{value}}`

> - Subscribe request (gnmic subscribe):
> 
> `gnmic -a {{ip:port}} --insecure subscribe --path {{path}}`

Reviewer: Going off of the [gNMI reference section 3.5](https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-specification.md#35-subscribing-to-telemetry-updates):

```suggestion
- Subscribe to target state updates under the subtree {{path}}:

`gnmic -a {{ip:port}} --insecure subscribe --path {{path}}`
```

# gnmic

> A gNMI command-line client.
> Manage gNMI network device configuration and view operational data.

> > More information: <https://gnmic.kmrd.dev>.

Reviewer: We may want to draw attention that there are tldr pages for subcommands, but let's get a 👍 / 👎  some of the other tldr collaborators before proceeding with this change. Thoughts?

```suggestion
> Some subcommands such as `get`, `set`, and `subscribe` have their own usage documentation, accessible via `tldr gnmic-subcommand`.

> More information: <https://gnmic.kmrd.dev>.
```

- Request device capabilities:

> `gnmic --address {{ip:port}} --insecure capabilities`

Reviewer: Question: we use `--insecure` as a flag for all of these examples but don't use them in other tldr pages. Is it worthwhile to have a single example explaining why this flag would be used, and then drop the flag from other examples?

Author: I removed the `--insecure` flag mentions.

- Request device capabilities:

> `gnmic --address {{ip:port}} capabilities`

Reviewer 2: IMO, it's better to provide a specific ip and port examples because `--address` is a long option and describes the meaning of the following placeholder.

Author: I think 127.0.0.1 is the only address that wouldn't work and addresses should be formatted as `address_of_server:port_to_connect`

- Provide a username and password to fetch device capabilities:

> `gnmic --address {{ip:port}} --username {{username}} --password {{password}} capabilities`

Reviewer 2: ```suggestion
`gnmic --address {{ip:port}} --username {{value}} --password {{value}} capabilities`
```

It's better not to repeat when possible.

Author: Sounds good

Reviewer 2: There is formal rule for this now but it should be. At the end of the day I can mark this conversation as resolved due to the reason mentioned earlier.

Reviewer 3: When has this been added as a rule?

We should always use informative placeholders

Reviewer 2: I've missed "no".

- Get a snapshot of the device state at a specific path:

> `gnmic -a {{ip:port}} get --path {{path}}`

Reviewer 2: ```suggestion
`gnmic -a {{ip:port}} get --path {{/path/to/device}}`
```

I am doubt about how to write this.
